### PR TITLE
External storage tables are now created in a migration

### LINKED ISSUE #26907: Upgrade from 9.1.3 to master :bomb:
### Steps:

1. Setup v9.1.3 from git
1. Checkout master (5a10d04488afd574dc2f8558cfd0ff4a8e204ddc) and run `make clean; make`
1. Run `occ upgrade` from the CLI

### Expected
Upgrade success.

### Actual
Upgrade failed:
```
Set log level to debug
Turned on maintenance mode
Updating database schema
Updated database
Disabled 3rd-party app: admin_migrate
Disabled 3rd-party app: files_odfviewer
Disabled incompatible app: files_texteditor
Disabled incompatible app: files_videoviewer
Disabled 3rd-party app: files_videoviewer
Disabled 3rd-party app: user_migrate
Drop old database tables

 Done
 28/28 [============================] 100%
Remove old (< 9.0) calendar/contact shares
 Done
 4/4 [============================] 100%
Fix permissions so avatars can be stored again
 Done
 2/2 [============================] 100%
Move user avatars outside the homes to the new location
 Starting ...
 0/1 [>---------------------------]   0%Doctrine\DBAL\Exception\TableNotFoundException: An exception occurred while executing 'SELECT m.`mount_id`, `mount_point`, `storage_backend`, `auth_backend`, `priority`, m.`type` FROM `oc_external_mounts` m INNER JOIN `oc_external_applicable` a ON m.`mount_id` = a.`mount_id` WHERE (a.`type` = ?) AND (a.`value` = ?) AND (m.`type` = '1')' with params [3, "admin"]:

SQLSTATE[42S02]: Base table or view not found: 1146 Table 'owncloud.oc_external_mounts' doesn't exist
Update failed
Maintenance mode is kept active
Reset log level
```

### Log
```
{"reqId":"tAyO1k42KrQVZ0rHQNAC","remoteAddr":"","app":"core","message":"Exception: {\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\TableNotFoundException\",\"Message\":\"An exception occurred while executing 'SELECT m.`mount_id`, `mount_point`, `storage_backend`, `auth_backend`, `priority`, m.`type` FROM `oc_external_mounts` m INNER JOIN `oc_external_applicable` a ON m.`mount_id` = a.`mount_id` WHERE (a.`type` = ?) AND (a.`value` = ?) AND (m.`type` = '1')' with params [3, \\\"admin\\\"]:\\n\\nSQLSTATE[42S02]: Base table or view not found: 1146 Table 'owncloud.oc_external_mounts' doesn't exist\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(128): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Connection.php(836): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'SELECT m.`mount...', Array)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/DB\\\/Connection.php(188): Doctrine\\\\DBAL\\\\Connection->executeQuery('SELECT m.`mount...', Array, Array, NULL)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Query\\\/QueryBuilder.php(206): OC\\\\DB\\\\Connection->executeQuery('SELECT m.`mount...', Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/DB\\\/QueryBuilder\\\/QueryBuilder.php(141): Doctrine\\\\DBAL\\\\Query\\\\QueryBuilder->execute()\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/DBConfigService.php(335): OC\\\\DB\\\\QueryBuilder\\\\QueryBuilder->execute()\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/DBConfigService.php(145): OC\\\\Files\\\\External\\\\Service\\\\DBConfigService->getMountsFromQuery(Object(OC\\\\DB\\\\QueryBuilder\\\\QueryBuilder))\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/UserGlobalStoragesService.php(76): OC\\\\Files\\\\External\\\\Service\\\\DBConfigService->getAdminMountsFor(3, 'admin')\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/StoragesService.php(128): OC\\\\Files\\\\External\\\\Service\\\\UserGlobalStoragesService->readDBConfig()\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/StoragesService.php(178): OC\\\\Files\\\\External\\\\Service\\\\StoragesService->readConfig()\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/StoragesService.php(187): OC\\\\Files\\\\External\\\\Service\\\\StoragesService->getAllStorages()\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/Service\\\/UserGlobalStoragesService.php(109): OC\\\\Files\\\\External\\\\Service\\\\StoragesService->getStorages()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/External\\\/ConfigAdapter.php(119): OC\\\\Files\\\\External\\\\Service\\\\UserGlobalStoragesService->getUniqueStorages()\\n#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Config\\\/MountProviderCollection.php(76): OC\\\\Files\\\\External\\\\ConfigAdapter->getMountsForUser(Object(OC\\\\User\\\\User), Object(OC\\\\Files\\\\Storage\\\\StorageFactory))\\n#14 [internal function]: OC\\\\Files\\\\Config\\\\MountProviderCollection->OC\\\\Files\\\\Config\\\\{closure}(Object(OC\\\\Files\\\\External\\\\ConfigAdapter))\\n#15 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Config\\\/MountProviderCollection.php(77): array_map(Object(Closure), Array)\\n#16 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Filesystem.php(443): OC\\\\Files\\\\Config\\\\MountProviderCollection->getMountsForUser(Object(OC\\\\User\\\\User))\\n#17 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Repair\\\/MoveAvatarOutsideHome.php(106): OC\\\\Files\\\\Filesystem::initMountPoints('admin')\\n#18 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Repair\\\/MoveAvatarOutsideHome.php(134): OC\\\\Repair\\\\MoveAvatarOutsideHome->moveAvatars(Object(OC\\\\Repair), Object(OC\\\\User\\\\User))\\n#19 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/User\\\/Manager.php(411): OC\\\\Repair\\\\MoveAvatarOutsideHome->OC\\\\Repair\\\\{closure}(Object(OC\\\\User\\\\User))\\n#20 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Repair\\\/MoveAvatarOutsideHome.php(140): OC\\\\User\\\\Manager->callForSeenUsers(Object(Closure))\\n#21 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Repair.php(89): OC\\\\Repair\\\\MoveAvatarOutsideHome->run(Object(OC\\\\Repair))\\n#22 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Updater.php(254): OC\\\\Repair->run()\\n#23 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Updater.php(121): OC\\\\Updater->doUpgrade('9.2.0.3', '9.1.2.5')\\n#24 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/core\\\/Command\\\/Upgrade.php(259): OC\\\\Updater->upgrade()\\n#25 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/symfony\\\/console\\\/Command\\\/Command.php(256): OC\\\\Core\\\\Command\\\\Upgrade->execute(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#26 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(818): Symfony\\\\Component\\\\Console\\\\Command\\\\Command->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#27 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(186): Symfony\\\\Component\\\\Console\\\\Application->doRunCommand(Object(OC\\\\Core\\\\Command\\\\Upgrade), Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#28 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(117): Symfony\\\\Component\\\\Console\\\\Application->doRun(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#29 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Console\\\/Application.php(158): Symfony\\\\Component\\\\Console\\\\Application->run(NULL, NULL)\\n#30 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/console.php(99): OC\\\\Console\\\\Application->run()\\n#31 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/occ(11): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#32 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":53}","level":3,"time":"2017-01-09T10:21:04+00:00","method":"--","url":"--","user":"--"}
```

Seems to be a clash between the "move avatars outside repair code" and the "move files_external to core" code. For some reason the table might not exist yet at the time this repair step is executed. Need to check why...

### PR DESCRIPTION
**Pull request by PVince81**
## Description
Reimplement the old files_external DB structure as a Doctrine migration.

## Related Issue
Fixes https://github.com/owncloud/core/issues/26907

## Motivation and Context
Because starting with 10.0 db_structure.xml is not used any more on updates and we need those tables.

## How Has This Been Tested?
- [x] TEST: setup this branch from scratch: tables properly created
- [x] TEST: setup OC 9.1 without files_external, migrate to this branch: tables properly created
- [x] TEST: setup OC 9.1 with files_external enabled and a configured ext storage, migrate to this branch: existing tables were kept with contents

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## TODOs:
- [x] ⚠️ make sure the table structure is exactly the same as it was when created from db_structure

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] BUG: wrong integer type
- [ ] BUG: wrong collation

left is 9.1, right is this branch:
```
<   `applicable_id` bigint(20) NOT NULL AUTO_INCREMENT,
<   `mount_id` bigint(20) NOT NULL,
---
>   `applicable_id` int(11) NOT NULL AUTO_INCREMENT,
>   `mount_id` int(11) NOT NULL,
```

**Comment by PVince81** [comment]
Added explicit collation and explicit bigint(20) values.
I hope this is correct...

I think I shouldn't need to explicitly specify the collation, see https://github.com/owncloud/core/issues/26925

Also I'm a bit surprised that the original schema had 256 chars for "mountpoint" specified, but the resulting schema in Mysql only had 128 chars defined. That's why I also set it to 128 here in the migration...

Basically all I tried to do is to prevent any differences between the 9.1 schema and the current one by diffing the schema dumps.

@DeepDiver1975 @butonic please review

**Comment by PVince81** [comment]
```
    Test\Files\External\Service\DBConfigServiceTest.testAddApplicable
    Test\Files\External\Service\DBConfigServiceTest.testRemoveApplicableGlobal
    Test\Files\External\Service\DBConfigServiceTest.testGetAdminMountsForGlobal
    Test\Files\External\Service\GlobalStoragesServiceTest::testAddStorage.testAddStorage with data set #0
    Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage.testUpdateStorage with data set #0
    Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage.testUpdateStorage with data set #1
    Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage.testUpdateStorage with data set #2
    Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage.testUpdateStorage with data set #3
    Test\Files\External\Service\GlobalStoragesServiceTest::testHooksAddStorage.testHooksAddStorage with data set #0
    Test\Files\External\Service\GlobalStoragesServiceTest::testHooksUpdateStorage.testHooksUpdateStorage with data set #0
    Test\Files\External\Service\GlobalStoragesServiceTest::testHooksUpdateStorage.testHooksUpdateStorage with data set #3
    Test\Files\External\Service\GlobalStoragesServiceTest::testHooksDeleteStorage.testHooksDeleteStorage with data set #1
    Test\Files\External\Service\GlobalStoragesServiceTest::testDeleteStorage.testDeleteStorage with data set #0
    Test\Files\External\Service\GlobalStoragesServiceTest::testDeleteStorage.testDeleteStorage with data set #1
    Test\Files\External\Service\GlobalStoragesServiceTest.testGetStoragesBackendNotVisible
    Test\Files\External\Service\GlobalStoragesServiceTest.testGetStoragesAuthMechanismNotVisible
    Test\Files\External\Service\GlobalStoragesServiceTest.testUpdateStorageMountPoint
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetStorageWithApplicable.testGetStorageWithApplicable with data set #0
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testDeleteStorage.testDeleteStorage with data set #0
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testDeleteStorage.testDeleteStorage with data set #1
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #0
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #1
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #2
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #3
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #8
    Test\Files\External\Service\UserGlobalStoragesServiceTest::testGetUniqueStorages.testGetUniqueStorages with data set #9
    Test\Files\External\Service\UserGlobalStoragesServiceTest.testUpdateStorage
```

Looks like I messed up some indices:
```
Integrity constraint violation: 19 NOT NULL constraint failed: oc_external_applicable.value
```

**Comment by PVince81** [comment]
Either Oracle had a hiccup or the migrations stuff doesn't work properly with Oracle:
```
19:32:57   [Doctrine\DBAL\Exception\TableNotFoundException]                                                                                                                                                                                                                                                                              
19:32:57   An exception occurred while executing 'SELECT m."mount_id", "mount_point", "storage_backend", "auth_backend", "priority", m."type" FROM "oc_external_mounts" m INNER JOIN "oc_external_applicable" a ON m."mount_id" = a."mount_id" WHERE (a."type" = ?) AND (a."value" = ?) AND (m."type" = 1)' with params [3, "admin23"]:  
19:32:57   ORA-00942: table or view does not exist
```

**Comment by PVince81** [comment]
Rebased + squashed. Let's hope the Oracle tests work now.

File: core/Migrations/Version20170111103310.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+/**
+ * Initial DB creation for external storages
+ */
+class Version20170111103310 extends AbstractMigration {
+	/**
+	 * @param Schema $schema
+	 */
+	public function up(Schema $schema) {
+		$prefix = $this->connection->getPrefix();
```
Comment (PVince81) [inline core/Migrations/Version20170111103310.php]: No, it's the ownCloud connection. Else this PR would not work at all at setup time. Try it 😉

**Comment by PVince81** [comment]
:+1: for @DeepDiver1975's fix

**Comment by PVince81** [comment]
@DeepDiver1975 @VicDeo is my commit acceptable then and can we merge it ?

File: core/Migrations/Version20170111103310.php (hunk 1 above)
Comment (VicDeo) [inline core/Migrations/Version20170111103310.php]: @PVince81 I was kicked in the a** by @DeepDiver1975 several times for exactly that kind of magic

Comment (DeepDiver1975) [inline core/Migrations/Version20170111103310.php]: I'm very aware of this - thanks for fighting me with my own weapons :hocho: .... maybe we need to come up with our own AbstractMigration class to allow proper handling of things like this .... let's keep this out of this pr for now.

**Comment by PVince81** [comment]
@DeepDiver1975 looked into the Oracle problem last week and said we'll need our own wrapper classes to Oraclify whatever input is coming in, like for example backtick-escaping stuff.

**Comment by PVince81** [comment]
Superseded by https://github.com/owncloud/core/pull/27041
